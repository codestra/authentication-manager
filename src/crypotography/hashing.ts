import { createHmac, Hmac, randomBytes } from 'crypto';

/**
 * generates random string of characters i.e salt
 * @function
 * @param {number} length - Length of the random string.
 */
function genRandomString(length: number): string {
  return randomBytes(Math.ceil(length / 2))
    .toString('hex') /** convert to hexadecimal format */
    .slice(0, length); /** return required number of characters */
}

/**
 * hash password with sha512.
 * @function
 * @param {string} password - List of required fields.
 * @param {string} salt - Data to be validated.
 */
export function createHash(data: any): { salt: string; password: string } {
  const salt: string = genRandomString(16); /** Gives us salt of length 16 */
  const hash: Hmac = createHmac('sha512', salt); /** Hashing algorithm sha512 */
  hash.update(data);
  const value: string = hash.digest('hex');
  return {
    salt,
    password: value,
  };
}

export function validateHash(data: any, salt: string): { password: string } {
  const hash: Hmac = createHmac('sha512', salt); /** Hashing algorithm sha512 */
  hash.update(data);
  const value: string = hash.digest('hex');
  return {
    password: value,
  };
}
