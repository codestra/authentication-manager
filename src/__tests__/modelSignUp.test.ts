import { MongoMemoryServer } from 'mongodb-memory-server';
import mongoose from 'mongoose';

process.env = { PASSWORD_RESET_EXPIRES: '8640000', JWT_EXPIRES: '9999y', JWT_SECRET: '123123' };
import { modelSignUp } from '../index';
import MockModel from '../mocks/model';

import * as cryptography from '../crypotography/hashing';

describe('modelSignUp', () => {
  let mongoServer: MongoMemoryServer;

  beforeAll(async () => {
    mongoServer = await MongoMemoryServer.create();
    await mongoose.connect(mongoServer.getUri(), { useNewUrlParser: true, useUnifiedTopology: true });
  });

  afterAll(async () => {
    if (mongoServer) {
      await mongoServer.stop();
      await mongoose.disconnect();
    }
  });

  it('should sign up the model', async () => {
    const result = await modelSignUp({
      Model: MockModel,
      variables: { email: 'sajad.ghawami@codestra.io', password: '123123' },
    });

    const data = await MockModel.findOne();
    expect(data._id.toString()).toBe(result._id);
  });

  it('should throw an duplicate key error', async () => {
    await expect(async () => {
      await modelSignUp({
        Model: MockModel,
        variables: { email: 'sajad.ghawami@codestra.io', password: '123123' },
      });
    }).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should throw an error', async () => {
    jest.spyOn(cryptography, 'createHash').mockImplementationOnce(() => {
      throw new Error('MOCKED_ERROR');
    });

    await expect(async () => {
      await modelSignUp({
        Model: MockModel,
        variables: { email: 'sajad.ghawami2@codestra.io', password: '123123' },
      });
    }).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should call the onComplete callback', async () => {
    const onCompleted = jest.fn();

    await modelSignUp({
      Model: MockModel,
      variables: { email: 'sajad.ghawami1@codestra.io', password: '123123' },
      onCompleted,
    });

    expect(onCompleted).toBeCalled();
  });
});
