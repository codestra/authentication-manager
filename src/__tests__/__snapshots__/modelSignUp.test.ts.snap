// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modelSignUp should throw an duplicate key error 1`] = `"MongoError: E11000 duplicate key error dup key: { : \\"sajad.ghawami@codestra.io\\" }"`;

exports[`modelSignUp should throw an error 1`] = `"Error: MOCKED_ERROR"`;
